expression does not work on attribute with reference

### Version
```
Kestrel v1.5.7
```

### Data
```
tests/test_bundle_nt_1.json
```

### Huntflow
```
conns = GET network-traffic
        FROM file:///tmp/kestrel-test-data/nt_1.json
        WHERE src_ref.value LIKE '192.%'
        START 2022-01-01T00:00:00.000Z STOP 2023-01-01T00:00:00.000Z

c2 = conns WHERE dst_ref.value = '192.168.56.112'
```

### Error
```
[ERROR] MissingEntityAttribute: variable "conns.dst_ref" does not have required attribute "value"
remove transform or specify different variable in the Kestrel command.
```

--- a/src/kestrel/codegen/commands.py
+++ b/src/kestrel/codegen/commands.py
@@ -26,7 +26,7 @@ from copy import deepcopy
 
 from firepit.deref import auto_deref
 from firepit.exceptions import InvalidAttr
-from firepit.query import Limit, Offset, Order, Projection, Query
+from firepit.query import Limit, Offset, Order, Predicate, Projection, Query
 from firepit.stix20 import summarize_pattern
 
 from kestrel.semantics.reference import make_deref_func, make_var_timerange_func
@@ -657,6 +657,8 @@ def _filter_prefetched_process(
 def _set_projection(store, entity_table, query, paths):
     joins, proj = auto_deref(store, entity_table, paths=paths)
     query.joins.extend(joins)
+    joined = [j.name for j in query.joins]
+    _logger.debug("%s: joining %s", query.table.name, joined)
     if query.proj:
         # Need to merge projections?  More-specific overrides less-specific ("*")
         new_cols = []
@@ -674,11 +676,38 @@ def _set_projection(store, entity_table, query, paths):
         query.proj = proj
 
 
+def _get_pred_columns(preds: list):
+    for pred in preds:
+        if isinstance(pred.lhs, Predicate):
+            yield from _get_pred_columns([pred.lhs])
+            if isinstance(pred.rhs, Predicate):
+                yield from _get_pred_columns([pred.rhs])
+        else:
+            yield pred.lhs
+
+
+def _get_filt_columns(filts: list):
+    for filt in filts:
+        yield from _get_pred_columns(filt.preds)
+
+
 def _build_query(store, entity_table, qry, stmt):
     where = stmt.get("where")
     if where:
-        where.set_table(entity_table)
-        qry.append(where)
+        if isinstance(where, Query):
+            for j in where.joins:
+                _logger.debug("Anchoring JOIN to %s", qry.table.name)
+                j.prev_name = qry.table.name
+            qry.joins.extend(where.joins)
+            for col in _get_filt_columns(where.where):
+                if col.table is None:
+                    # Need to disambiguate any Predicate columns
+                    _logger.debug("Disambiguating predicate for %s", qry.table.name)
+                    col.table = qry.table.name
+            qry.where.extend(where.where)
+        else:
+            where.set_table(entity_table)
+            qry.append(where)
     attrs = stmt.get("attrs", "*")
     cols = attrs.split(",")
     _set_projection(store, entity_table, qry, cols)

--- a/src/kestrel/syntax/ecgpattern.py
+++ b/src/kestrel/syntax/ecgpattern.py
@@ -7,7 +7,8 @@ from typeguard import typechecked
 from typing import Tuple, Optional
 import datetime
 from abc import ABC, abstractmethod
-from firepit.query import Filter, Predicate
+from firepit.query import Column, Filter, Predicate, Query
+from firepit.sqlstorage import get_path_joins
 from firepit.timestamp import timefmt
 from kestrel.syntax.utils import merge_timeranges
 from kestrel.syntax.reference import (
@@ -20,6 +21,15 @@ from kestrel.exceptions import (
     InvalidECGPattern,
     KestrelInternalError,
 )
+
+
+def _merge_queries(lhs: Query, op: str, rhs: Query):
+    result = Query(lhs.table)
+    result.joins = lhs.joins + rhs.joins
+    lpred = lhs.where[0].preds[0]
+    rpred = rhs.where[0].preds[0]
+    result.where.append(Filter([Predicate(lpred, op, rpred)]))
+    return result
 
 
 class ExtCenteredGraphConstruct(ABC):
@@ -115,7 +125,7 @@ class ExtCenteredGraphPattern(ExtCenteredGraphConstruct):
         if self.graph is None:
             return None
         else:
-            return Filter([self.graph.to_firepit(self.center_entity_type)])
+            return self.graph.to_firepit(self.center_entity_type)
 
     def deref(self, deref_func, get_timerange_func):
         if self.graph is not None:
@@ -193,7 +203,7 @@ class ECGPJunction(ExtCenteredGraphConstruct):
         )
 
     def to_firepit(self, center_entity_type: str):
-        return Predicate(
+        return _merge_queries(
             self.lhs.to_firepit(center_entity_type),
             self.relation,
             self.rhs.to_firepit(center_entity_type),
@@ -240,7 +250,17 @@ class ECGPComparison(ExtCenteredGraphConstruct):
     def to_firepit(self, center_entity_type: str):
         if self.etype and self.etype != center_entity_type:
             raise KestrelNotImplemented("firepit filtering on linked entities")
-        return Predicate(self.attribute, self.op, self.value)
+        joins, target_type, target_attr = get_path_joins(
+            None, self.etype, self.attribute
+        )
+        if target_type:
+            attribute = Column(target_attr, table=target_type)
+        else:
+            attribute = self.attribute
+        qry = Query()
+        qry.joins = joins
+        qry.where = [Filter([Predicate(attribute, self.op, self.value)])]
+        return qry
 
     def deref(self, deref_func, get_timerange_func):
         xs, tr = deref_and_flatten_value_to_list(

--- a/tests/test_session.py
+++ b/tests/test_session.py
@@ -310,3 +310,40 @@ language:
     full_path = os.path.join(os.getcwd(), config_name)
     assert os.environ[kestrel.config.CONFIG_PATH_ENV_VAR] == full_path 
     assert os.environ[kestrel_datasource_stixshifter.config.PROFILE_PATH_ENV_VAR] == full_path
+
+
+def test_where_deref_network_traffic(fake_bundle_file):
+    '''#290: expression does not work on attribute with reference'''
+    with Session(debug_mode=True) as session:
+        execute(
+            session,
+            f"""conns = get network-traffic
+            from file://{fake_bundle_file}
+            where [network-traffic:dst_port < 10000]""",
+        )
+        conns = get_df(session, "conns")
+        assert len(conns.index) == 100
+        execute(session, "c2 = conns where src_ref.value = '192.168.212.97'")
+        c2 = session.get_variable("c2")
+        assert len(c2) == 1
+        assert c2[0]["dst_port"] == 22
+
+        execute(session, "c2 = conns where src_ref.value = '192.168.121.121' and dst_ref.value = '10.0.0.134'")
+        c2 = session.get_variable("c2")
+        assert len(c2) == 1
+        assert c2[0]["dst_port"] == 3128
+
+
+def test_where_deref_process(cbcloud_powershell_bundle):
+    '''#290: expression does not work on attribute with reference'''
+    with Session() as session:
+        script = (
+            "x = get process"
+            f" from file://{cbcloud_powershell_bundle}"
+            " where [process:name = 'powershell.exe']"
+        )
+        execute(session, script)
+        execute(session, "y = x where parent_ref.pid = 1544")
+        y = session.get_variable("y")
+        assert len(y) == 1
+        assert y[0]["parent_ref.x_unique_id"] == "MYORGIDX-02629f16-00000608-00000000-1d71d10a09cc7c4"

